Realize hotel reservation system

behumb opened the pull request:
- Add hotel system classes: Hotel, Room, Person
 - Add work visualization in main() main.py

behumb committed:
Realize hotel reservation system

 - Add hotel system classes: Hotel, Room, Person
 - Add work visualization in main() main.py

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,98 @@
+
+    @money.setter
+    def money(self, sum):
+        self.__money += sum
```
Why it's not `self.__money = sum`?

denisorehovsky commented on hotel_reservation_system/main.py:
```diff
@@ -0,0 +1,22 @@
+from hotel_system import *
```
You should never use `from x import *`. Always import explicitly. https://www.geeksforgeeks.org/why-import-star-in-python-is-a-bad-idea/#:~:text=Using%20import%20*%20in%20python%20programs,other%20libraries%20that%20you%20import.

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,98 @@
+        self.__tenant = None
+
+    def is_booked(self):
+        return not self.tenant is None
```
`self.tenant is not None` is more readable

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,98 @@
+        self.rooms.extend(rooms)
+
+    def release_room(self, room):
+        room.tenant = None
```
What if you have:
- Hotel A with Room 1 and Room 2
- Hotel B with Room 10 and Room 20

then you call:
hotel_a.release_room(room_10)

This code will work even though room_10 is not a part of hotel_a. It's not correct.

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,98 @@
+            person.money -= days_count * room.cost
+            print('Congratulations! Room is booked')
+        else:
+            print('You can not book this room(')
```
It's much better to raise custom exceptions. For example, you can create:
```
class RoomIsNotBookableError(Exception):
    pass
```

and then raise it in your code:
```
raise RoomIsNotBookableError()
```

print isn't useful here. It doesn't help you. What if you want to stop your program if your room is not bookable?

denisorehovsky commented on hotel_reservation_system/main.py:
```diff
@@ -0,0 +1,22 @@
+    hotel.show_price_list()
+    room = hotel.get_room_by_number(3)
+    room_2 = hotel.get_room_by_number(3)
+    hotel.allocate_room(person_1, room, 2)
```
When I read this line, I don't understand what is 2. I need to open a class.

This is more readable:
```
hotel.allocate_room(person=person_1, room=room, days_count=2)
```

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,98 @@
+        return None
+
+    def allocate_room(self, person, room, days_count):
+        if not room.tenant and days_count * person.money:
```
`days_count * person.money` - what does it check?

denisorehovsky replied:
Also, I don't think that having `days_count` is a good idea. What if I allocated room for 1 day, but didn't call `release_room` after that? Your room will not be bookable.

Also, if I run the program today and set days_count = 2 and then save it to the database. Then if I run this program tomorrow, days_count will still be 2. How do I know when to release a room?

I would use dates instead.

behumb committed:
Fix problems

 - Replace days_count to two datetime objects date_start and date_end
 - Add field Room.release_date for checking room status
 - Fix problem with allocate room

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,124 @@
+        self.__release_date = None
+
+    def is_free(self):
+        if self.release_date is None:
```
It's shorter to write:
```
return self.release_date is None or self.release_date < datetime.now()
```

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,124 @@
+
+    def allocate_room(self, person, room, date_start, date_end):
+        days_count = (date_end - date_start).days
+        if days_count * room.cost < person.money and room in self.rooms and room.is_free():
```
I think it's more correct:
```
days_count * room.cost <= person.money
```

denisorehovsky replied:
Also, I think it's better to create a function for this condition:
```
def can_allocate_room(self, person...):
    return days_count * room.cost < person.money and room in self.rooms and room.is_free()

def allocate_room(self, person...):
   if not self.can_allocate_room(person,...):
       raise RoomIsNotBookableError
    ....
```

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,124 @@
+            person.money -= days_count * room.cost
+            print('Congratulations! Room is booked')
+        else:
+            raise RoomIsNotBookableError('You can not book this room(')
```
This message I think is better:
```
f'You can not book romm #{room.number}'
```

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,124 @@
+    pass
+
+
+class RoomIsNotExistInHotelError(Exception):
```
RoomIsNotExistInHotelError -> RoomDoesNotExistInHotelError

denisorehovsky reviewed (changes requested):
Imagine the case:
Person A booked Room 1 from February 11 to February 13

How Person B can book Room 1 from February 1 to February 5?

How can I do that using your classes?

behumb committed:
Fix problem with booking

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,163 @@
+
+class RoomDoesNotExistInHotelError(Exception):
+    pass
+
```
Should be two spaces between classes

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,163 @@
+
+    def is_free(self, date_start, date_end):
+        for order in self.orders:
+            if self.is_date_between(date_start, order.date_start, order.date_end) or self.is_date_between(date_end,
```
What if you have an order from Feb 5 to Feb 7. The user is trying to create an order from Feb 1 to Feb 10. I think this check is not going to work.

denisorehovsky replied:
Also, the line is too long. You should split lines if they are long. For example
```
if (
    self.is_date_between(date_start, order.date_start, order.date_end)
    or self.is_date_between(date_end, ...
):
    123
```

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,163 @@
+
+    def release_room(self, room, order):
+        if room in self.rooms:
+            room.orders.remove(order)
```
What if the room doesn't have the order? The code will raise ValueError

behumb committed:
Fix booking error and error with release room

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,167 @@
+            raise OrderDoesNotExistError()
+    
+
+
```
Should be two spaces, not three between classes. Install flake8 or something. You should have red alerts in your editor

denisorehovsky commented on hotel_reservation_system/hotel_system.py:
```diff
@@ -0,0 +1,167 @@
+    def is_date_between(self, date, date_start, date_end):
+        return date_start <= date <= date_end
+
+    def is_free(self, date_start, date_end):
```
This function still doesn't work correctly. If there is an order from Feb 3 to Feb 6 and you want to book from Feb 6 to Feb 10, it's not going to work.

But never mind.

behumb committed:
Fix another one problem with definition of free date for booking
